File: src/bin/test_examples.rs
use std::fs;
use std::io;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::exit;
use std::process::{Command, Stdio};

fn get_subdirs(dir: &str) -> Result<Vec<String>, io::Error> {
    let path = Path::new(dir);
    let mut subdirs = Vec::new();
    let entries = fs::read_dir(path)?;
    for entry_result in entries {
        let entry = entry_result?;
        let path = entry.path();
        if path.is_dir() {
            // Get the directory name as a string
            if let Some(dir_name) = path.file_name() {
                if let Some(dir_str) = dir_name.to_str() {
                    subdirs.push(dir_str.to_string());
                } else {
                    // Handle non-Unicode directory names
                    return Err(io::Error::new(
                        io::ErrorKind::InvalidData,
                        format!("Directory name contains invalid Unicode: {:?}", dir_name),
                    ));
                }
            } else {
                // Handle paths with no file name component
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!("Failed to get directory name from path: {:?}", path),
                ));
            }
        }
    }
    Ok(subdirs)
}

fn main() {
    // Get a list of the example programs
    let dirs = match get_subdirs("examples") {
        Ok(dirs) => dirs,
        Err(e) => {
            println!("Error reading subdirectories: {}", e);
            std::process::exit(1);
        }
    };

    let mut passed_count = 0;
    let mut skipped_count = 0;

    // For each example program
    for name in dirs {
        // If output.txt exists, use it as the basis for comparison
        let expected_output_file = PathBuf::from("examples").join(&name).join("output.txt");
        if !expected_output_file.exists() {
            skipped_count += 1;
            continue;
        }
        let expected_output_file = expected_output_file
            .into_os_string()
            .into_string()
            .expect("Path contains invalid UTF-8");
        let expected_output = match fs::read_to_string(&expected_output_file) {
            Ok(content) => content,
            Err(e) => {
                println!("Failed to read {}: {}", expected_output_file, e);
                exit(1);
            }
        };

        // If input.txt exists, use it for input
        let input_file = PathBuf::from("examples").join(&name).join("input.txt");
        let input = if input_file.exists() {
            let input_file = input_file
                .into_os_string()
                .into_string()
                .expect("Path contains invalid UTF-8");
            match fs::read_to_string(&input_file) {
                Ok(content) => content,
                Err(e) => {
                    println!("Failed to read {}: {}", input_file, e);
                    exit(1);
                }
            }
        } else {
            "".to_string()
        };

        // Run the program
        let program_file = PathBuf::from("examples")
            .join(&name)
            .join(format!("{}.va", name))
            .into_os_string()
            .into_string()
            .expect("Path contains invalid UTF-8");

        // Create a command with piped stdin
        let mut cmd = Command::new("./target/debug/verbena");
        cmd.arg(&program_file)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());

        // Spawn the command
        let mut child = match cmd.spawn() {
            Ok(child) => child,
            Err(e) => {
                println!("{}", program_file);
                println!("Failed to run interpreter: {}", e);
                exit(1);
            }
        };

        // Write the input to stdin
        if !input.is_empty() {
            if let Some(stdin) = child.stdin.as_mut() {
                if let Err(e) = stdin.write_all(input.as_bytes()) {
                    println!("{}", program_file);
                    println!("Failed to write to stdin: {}", e);
                    exit(1);
                }
            }
        }

        // Wait for the command to complete and collect output
        let output = match child.wait_with_output() {
            Ok(output) => output,
            Err(e) => {
                println!("{}", program_file);
                println!("Failed to wait for interpreter: {}", e);
                exit(1);
            }
        };

        let actual_output = match String::from_utf8(output.stdout) {
            Ok(string) => string,
            Err(e) => {
                println!("{}", program_file);
                println!("Actual output not valid UTF-8: {}", e);
                exit(1);
            }
        };
        let stderr_output = match String::from_utf8(output.stderr) {
            Ok(string) => string,
            Err(e) => {
                println!("{}", program_file);
                println!("stderr output not valid UTF-8: {}", e);
                exit(1);
            }
        };
        let exit_code = output.status.code();

        // Error
        if !stderr_output.is_empty() {
            println!("{}", program_file);
            print!("{}", stderr_output);
            exit(1);
        }
        if exit_code != Some(0) {
            println!("{}", program_file);
            println!("Exit code {:?}", exit_code);
            exit(1);
        }

        // Compare outputs
        // Watch out for output.txt's with MS-DOS line endings!
        if actual_output == expected_output {
            passed_count += 1;
        } else {
            println!("{}", program_file);
            println!(
                "Output doesn't match expected.\nExpected:\n{}\nActual:\n{}",
                expected_output, actual_output
            );
        }
    }
    println!("Passed : {}", passed_count);
    println!("Skipped: {}", skipped_count);
}

File: test_examples.py
import os
import subprocess
import sys
from pathlib import Path


def get_example_files(directory: str) -> list[str]:
    """
    Get all Verbena (.va) files in the specified directory.
    Args:
        directory: The directory to search for .va files
    Returns:
        A list of file names without the .va extension
    Raises:
        OSError: If there's an error reading the directory
    """
    path = Path(directory)
    example_files = []
    try:
        for entry in os.listdir(path):
            if entry.endswith(".va"):
                # Remove the .va extension to get the base name
                example_files.append(entry[:-3])
    except OSError as e:
        raise OSError(f"Error reading directory: {e}")
    return example_files


def main():
    # Get a list of the example programs
    try:
        example_names = get_example_files("examples")
    except OSError as e:
        print(e)
        sys.exit(1)

    passed_count = 0
    skipped_count = 0

    # For each example program
    for name in example_names:
        # Check if corresponding output file exists
        expected_output_file = Path("examples_output") / f"{name}.txt"
        if not expected_output_file.exists():
            skipped_count += 1
            continue

        try:
            expected_output = expected_output_file.read_text()
        except OSError as e:
            print(f"Failed to read {expected_output_file}: {e}")
            sys.exit(1)

        # Get the program file path
        program_file = Path("examples") / f"{name}.va"

        # First, compile the Verbena source to a.js
        try:
            compile_proc = subprocess.run(
                ["./target/debug/verbena", str(program_file)],
                capture_output=True,
                text=True,
            )

            if compile_proc.returncode != 0:
                print(f"{program_file}")
                print(f"Failed to compile: {compile_proc.stderr}")
                sys.exit(1)
        except Exception as e:
            print(f"{program_file}")
            print(f"Failed to run compiler: {e}")
            sys.exit(1)

        # Now run the compiled JavaScript with node
        try:
            proc = subprocess.Popen(
                ["node", "a.js"],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,  # This ensures string I/O instead of bytes
            )
        except Exception as e:
            print(f"{program_file}")
            print(f"Failed to run node on compiled output: {e}")
            sys.exit(1)

        # Write empty input to stdin
        try:
            stdout, stderr = proc.communicate(input="")
        except Exception as e:
            print(f"{program_file}")
            print(f"Failed during program execution: {e}")
            sys.exit(1)

        exit_code = proc.returncode

        # Error
        if stderr:
            print(f"{program_file}")
            print(stderr, end="")
            sys.exit(1)

        if exit_code != 0:
            print(f"{program_file}")
            print(f"Exit code {exit_code}")
            sys.exit(1)

        # Compare outputs
        if stdout == expected_output:
            passed_count += 1
        else:
            print(f"{program_file}")
            print(
                f"Output doesn't match expected.\nExpected:\n{expected_output}\nActual:\n{stdout}"
            )

    print(f"Passed : {passed_count}")
    print(f"Skipped: {skipped_count}")


if __name__ == "__main__":
    main()
